File: Backend/innkt.Officer/Services/SearchService.cs
using innkt.Common.Models;
using Microsoft.Extensions.Logging;
using System.Text.RegularExpressions;

namespace innkt.Officer.Services
{
    public class SearchService
    {
        private readonly ILogger<SearchService> _logger;
        private readonly TrendingAlgorithmService _trendingService;

        public SearchService(
            ILogger<SearchService> logger,
            TrendingAlgorithmService trendingService)
        {
            _logger = logger;
            _trendingService = trendingService;
        }

        /// <summary>
        /// Perform comprehensive search across all content types
        /// </summary>
        public async Task<SearchResult> SearchAsync(SearchRequest request)
        {
            try
            {
                var result = new SearchResult
                {
                    Query = request.Query,
                    TotalResults = 0,
                    SearchTime = DateTime.UtcNow
                };

                var searchTasks = new List<Task>();

                // Search posts if requested
                if (request.IncludePosts)
                {
                    searchTasks.Add(Task.Run(async () =>
                    {
                        result.Posts = await SearchPostsAsync(request);
                    }));
                }

                // Search users if requested
                if (request.IncludeUsers)
                {
                    searchTasks.Add(Task.Run(async () =>
                    {
                        result.Users = await SearchUsersAsync(request);
                    }));
                }

                // Search groups if requested
                if (request.IncludeGroups)
                {
                    searchTasks.Add(Task.Run(async () =>
                    {
                        result.Groups = await SearchGroupsAsync(request);
                    }));
                }

                // Search hashtags if requested
                if (request.IncludeHashtags)
                {
                    searchTasks.Add(Task.Run(async () =>
                    {
                        result.Hashtags = await SearchHashtagsAsync(request);
                    }));
                }

                // Wait for all searches to complete
                await Task.WhenAll(searchTasks);

                // Calculate total results
                result.TotalResults = result.Posts.Count + result.Users.Count + result.Groups.Count + result.Hashtags.Count;

                // Add search suggestions
                result.Suggestions = await GetSearchSuggestionsAsync(request.Query);

                _logger.LogInformation($"Search completed for query '{request.Query}' with {result.TotalResults} results");

                return result;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Error performing search for query '{request.Query}'");
                return new SearchResult
                {
                    Query = request.Query,
                    Error = "Search failed. Please try again."
                };
            }
        }

        /// <summary>
        /// Search posts with advanced filtering
        /// </summary>
        private async Task<List<SearchPostResult>> SearchPostsAsync(SearchRequest request)
        {
            try
            {
                // TODO: Implement actual database search
                // For now, return mock data
                var posts = new List<SearchPostResult>();

                // This would typically query your database with full-text search
                // Example: SELECT * FROM Posts WHERE CONTAINS(Content, @query) OR CONTAINS(Hashtags, @query)

                _logger.LogDebug($"Searching posts for query: {request.Query}");
                
                return posts;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error searching posts");
                return new List<SearchPostResult>();
            }
        }

        /// <summary>
        /// Search users with relevance scoring
        /// </summary>
        private async Task<List<SearchUserResult>> SearchUsersAsync(SearchRequest request)
        {
            try
            {
                // TODO: Implement actual database search
                var users = new List<SearchUserResult>();

                // This would typically query your database
                // Example: SELECT * FROM Users WHERE Username LIKE @query OR DisplayName LIKE @query OR Bio LIKE @query

                _logger.LogDebug($"Searching users for query: {request.Query}");
                
                return users;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error searching users");
                return new List<SearchUserResult>();
            }
        }

        /// <summary>
        /// Search groups with category filtering
        /// </summary>
        private async Task<List<SearchGroupResult>> SearchGroupsAsync(SearchRequest request)
        {
            try
            {
                // TODO: Implement actual database search
                var groups = new List<SearchGroupResult>();

                // This would typically query your database
                // Example: SELECT * FROM Groups WHERE Name LIKE @query OR Description LIKE @query

                _logger.LogDebug($"Searching groups for query: {request.Query}");
                
                return groups;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error searching groups");
                return new List<SearchGroupResult>();
            }
        }

        /// <summary>
        /// Search hashtags with usage statistics
        /// </summary>
        private async Task<List<SearchHashtagResult>> SearchHashtagsAsync(SearchRequest request)
        {
            try
            {
                // TODO: Implement actual database search
                var hashtags = new List<SearchHashtagResult>();

                // This would typically query your database
                // Example: SELECT Hashtag, COUNT(*) as UsageCount FROM Posts WHERE Hashtags LIKE @query GROUP BY Hashtag

                _logger.LogDebug($"Searching hashtags for query: {request.Query}");
                
                return hashtags;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error searching hashtags");
                return new List<SearchHashtagResult>();
            }
        }

        /// <summary>
        /// Get search suggestions based on query
        /// </summary>
        private async Task<List<string>> GetSearchSuggestionsAsync(string query)
        {
            try
            {
                if (string.IsNullOrWhiteSpace(query) || query.Length < 2)
                {
                    return new List<string>();
                }

                var suggestions = new List<string>();

                // TODO: Implement actual suggestion logic
                // This could come from:
                // - Popular searches
                // - User's search history
                // - Trending hashtags
                // - Common usernames

                _logger.LogDebug($"Getting search suggestions for query: {query}");
                
                return suggestions;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting search suggestions");
                return new List<string>();
            }
        }

        /// <summary>
        /// Get trending searches
        /// </summary>
        public async Task<List<TrendingSearch>> GetTrendingSearchesAsync(int limit = 10)
        {
            try
            {
                // TODO: Implement trending searches based on search frequency
                var trendingSearches = new List<TrendingSearch>();

                _logger.LogDebug($"Getting {limit} trending searches");
                
                return trendingSearches;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting trending searches");
                return new List<TrendingSearch>();
            }
        }

        /// <summary>
        /// Get search history for a user
        /// </summary>
        public async Task<List<SearchHistoryItem>> GetSearchHistoryAsync(string userId, int limit = 20)
        {
            try
            {
                // TODO: Implement search history storage and retrieval
                var history = new List<SearchHistoryItem>();

                _logger.LogDebug($"Getting search history for user {userId}");
                
                return history;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Error getting search history for user {userId}");
                return new List<SearchHistoryItem>();
            }
        }

        /// <summary>
        /// Save search query to user's history
        /// </summary>
        public async Task SaveSearchQueryAsync(string userId, string query, SearchResult result)
        {
            try
            {
                // TODO: Implement search history storage
                // This could be stored in Redis or database

                _logger.LogDebug($"Saving search query '{query}' for user {userId}");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Error saving search query for user {userId}");
            }
        }

        /// <summary>
        /// Clear search history for a user
        /// </summary>
        public async Task ClearSearchHistoryAsync(string userId)
        {
            try
            {
                // TODO: Implement search history clearing
                _logger.LogDebug($"Clearing search history for user {userId}");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Error clearing search history for user {userId}");
            }
        }

        /// <summary>
        /// Get search analytics
        /// </summary>
        public async Task<SearchAnalytics> GetSearchAnalyticsAsync(DateTime? fromDate = null, DateTime? toDate = null)
        {
            try
            {
                // TODO: Implement search analytics
                var analytics = new SearchAnalytics
                {
                    TotalSearches = 0,
                    UniqueQueries = 0,
                    MostSearchedTerms = new List<SearchTermStats>(),
                    SearchTrends = new List<SearchTrend>(),
                    Period = new DateRange
                    {
                        From = fromDate ?? DateTime.UtcNow.AddDays(-30),
                        To = toDate ?? DateTime.UtcNow
                    }
                };

                _logger.LogDebug("Getting search analytics");
                
                return analytics;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting search analytics");
                return new SearchAnalytics();
            }
        }

        /// <summary>
        /// Parse search query and extract filters
        /// </summary>
        private SearchQuery ParseSearchQuery(string query)
        {
            var parsedQuery = new SearchQuery
            {
                OriginalQuery = query,
                Terms = new List<string>(),
                Filters = new Dictionary<string, string>(),
                Hashtags = new List<string>(),
                Mentions = new List<string>()
            };

            if (string.IsNullOrWhiteSpace(query))
            {
                return parsedQuery;
            }

            // Extract hashtags
            var hashtagMatches = Regex.Matches(query, @"#(\w+)");
            foreach (Match match in hashtagMatches)
            {
                parsedQuery.Hashtags.Add(match.Groups[1].Value);
            }

            // Extract mentions
            var mentionMatches = Regex.Matches(query, @"@(\w+)");
            foreach (Match match in mentionMatches)
            {
                parsedQuery.Mentions.Add(match.Groups[1].Value);
            }

            // Extract filters (e.g., "type:post", "user:john", "date:today")
            var filterMatches = Regex.Matches(query, @"(\w+):(\w+)");
            foreach (Match match in filterMatches)
            {
                parsedQuery.Filters[match.Groups[1].Value] = match.Groups[2].Value;
            }

            // Extract remaining terms
            var cleanQuery = query;
            foreach (Match match in hashtagMatches.Cast<Match>().Concat(mentionMatches.Cast<Match>()).Concat(filterMatches.Cast<Match>()))
            {
                cleanQuery = cleanQuery.Replace(match.Value, "");
            }
            
            parsedQuery.Terms = cleanQuery.Split(' ', StringSplitOptions.RemoveEmptyEntries)
                .Where(term => !string.IsNullOrWhiteSpace(term))
                .ToList();

            return parsedQuery;
        }

        /// <summary>
        /// Calculate search relevance score
        /// </summary>
        private double CalculateRelevanceScore(string query, string content, Dictionary<string, double> weights)
        {
            if (string.IsNullOrWhiteSpace(query) || string.IsNullOrWhiteSpace(content))
            {
                return 0;
            }

            var queryTerms = query.ToLower().Split(' ', StringSplitOptions.RemoveEmptyEntries);
            var contentLower = content.ToLower();
            
            double score = 0;
            
            foreach (var term in queryTerms)
            {
                if (contentLower.Contains(term))
                {
                    // Exact match gets higher score
                    if (contentLower.Contains($" {term} ") || contentLower.StartsWith($"{term} ") || contentLower.EndsWith($" {term}"))
                    {
                        score += weights.GetValueOrDefault("exact_match", 1.0);
                    }
                    else
                    {
                        score += weights.GetValueOrDefault("partial_match", 0.5);
                    }
                }
            }

            return score;
        }
    }

    // Search Models
    public class SearchRequest
    {
        public string Query { get; set; } = string.Empty;
        public int Page { get; set; } = 1;
        public int PageSize { get; set; } = 20;
        public bool IncludePosts { get; set; } = true;
        public bool IncludeUsers { get; set; } = true;
        public bool IncludeGroups { get; set; } = true;
        public bool IncludeHashtags { get; set; } = true;
        public SearchFilters? Filters { get; set; }
        public SearchSort? Sort { get; set; }
    }

    public class SearchResult
    {
        public string Query { get; set; } = string.Empty;
        public int TotalResults { get; set; }
        public List<SearchPostResult> Posts { get; set; } = new();
        public List<SearchUserResult> Users { get; set; } = new();
        public List<SearchGroupResult> Groups { get; set; } = new();
        public List<SearchHashtagResult> Hashtags { get; set; } = new();
        public List<string> Suggestions { get; set; } = new();
        public DateTime SearchTime { get; set; }
        public string? Error { get; set; }
    }

    public class SearchPostResult
    {
        public string Id { get; set; } = string.Empty;
        public string Content { get; set; } = string.Empty;
        public string AuthorId { get; set; } = string.Empty;
        public string AuthorName { get; set; } = string.Empty;
        public string AuthorAvatar { get; set; } = string.Empty;
        public DateTime CreatedAt { get; set; }
        public int LikesCount { get; set; }
        public int CommentsCount { get; set; }
        public int SharesCount { get; set; }
        public List<string> Hashtags { get; set; } = new();
        public List<string> MediaUrls { get; set; } = new();
        public double RelevanceScore { get; set; }
        public string? HighlightedContent { get; set; }
    }

    public class SearchUserResult
    {
        public string Id { get; set; } = string.Empty;
        public string Username { get; set; } = string.Empty;
        public string DisplayName { get; set; } = string.Empty;
        public string Bio { get; set; } = string.Empty;
        public string AvatarUrl { get; set; } = string.Empty;
        public int FollowersCount { get; set; }
        public int FollowingCount { get; set; }
        public int PostsCount { get; set; }
        public bool IsVerified { get; set; }
        public bool IsFollowing { get; set; }
        public double RelevanceScore { get; set; }
    }

    public class SearchGroupResult
    {
        public string Id { get; set; } = string.Empty;
        public string Name { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public string Category { get; set; } = string.Empty;
        public string AvatarUrl { get; set; } = string.Empty;
        public string CoverUrl { get; set; } = string.Empty;
        public int MembersCount { get; set; }
        public int PostsCount { get; set; }
        public bool IsPrivate { get; set; }
        public bool IsJoined { get; set; }
        public double RelevanceScore { get; set; }
    }

    public class SearchHashtagResult
    {
        public string Hashtag { get; set; } = string.Empty;
        public int UsageCount { get; set; }
        public DateTime LastUsed { get; set; }
        public List<string> RecentPosts { get; set; } = new();
        public double RelevanceScore { get; set; }
    }

    public class SearchFilters
    {
        public DateTime? FromDate { get; set; }
        public DateTime? ToDate { get; set; }
        public List<string> Categories { get; set; } = new();
        public List<string> Types { get; set; } = new();
        public bool? HasMedia { get; set; }
        public int? MinLikes { get; set; }
        public int? MinComments { get; set; }
        public bool? IsVerified { get; set; }
        public bool? IsPrivate { get; set; }
    }

    public class SearchSort
    {
        public string Field { get; set; } = "relevance";
        public string Direction { get; set; } = "desc";
    }

    public class SearchQuery
    {
        public string OriginalQuery { get; set; } = string.Empty;
        public List<string> Terms { get; set; } = new();
        public Dictionary<string, string> Filters { get; set; } = new();
        public List<string> Hashtags { get; set; } = new();
        public List<string> Mentions { get; set; } = new();
    }

    public class TrendingSearch
    {
        public string Query { get; set; } = string.Empty;
        public int SearchCount { get; set; }
        public double TrendScore { get; set; }
        public DateTime LastSearched { get; set; }
    }

    public class SearchHistoryItem
    {
        public string Id { get; set; } = string.Empty;
        public string Query { get; set; } = string.Empty;
        public int ResultCount { get; set; }
        public DateTime SearchedAt { get; set; }
    }

    public class SearchAnalytics
    {
        public int TotalSearches { get; set; }
        public int UniqueQueries { get; set; }
        public List<SearchTermStats> MostSearchedTerms { get; set; } = new();
        public List<SearchTrend> SearchTrends { get; set; } = new();
        public DateRange Period { get; set; } = new();
    }

    public class SearchTermStats
    {
        public string Term { get; set; } = string.Empty;
        public int Count { get; set; }
        public double Trend { get; set; }
    }

    public class SearchTrend
    {
        public DateTime Date { get; set; }
        public int SearchCount { get; set; }
        public int UniqueQueries { get; set; }
    }

    public class DateRange
    {
        public DateTime From { get; set; }
        public DateTime To { get; set; }
    }
}

File: Frontend/innkt.react/src/services/search.service.ts
import { BaseApiService, socialApi, neurosparkApi } from './api.service';

export interface SearchRequest {
  query: string;
  page?: number;
  pageSize?: number;
  includePosts?: boolean;
  includeUsers?: boolean;
  includeGroups?: boolean;
  includeHashtags?: boolean;
  filters?: SearchFilters;
  sort?: SearchSort;
}

export interface SearchResult {
  query: string;
  totalResults: number;
  posts: SearchPostResult[];
  users: SearchUserResult[];
  groups: SearchGroupResult[];
  hashtags: SearchHashtagResult[];
  suggestions: string[];
  searchTime: string;
  error?: string;
}

export interface SearchPostResult {
  id: string;
  content: string;
  authorId: string;
  authorName: string;
  authorAvatar: string;
  createdAt: string;
  likesCount: number;
  commentsCount: number;
  sharesCount: number;
  hashtags: string[];
  mediaUrls: string[];
  relevanceScore: number;
  highlightedContent?: string;
}

export interface SearchUserResult {
  id: string;
  username: string;
  displayName: string;
  bio: string;
  avatarUrl: string;
  followersCount: number;
  followingCount: number;
  postsCount: number;
  isVerified: boolean;
  isFollowing: boolean;
  relevanceScore: number;
}

export interface SearchGroupResult {
  id: string;
  name: string;
  description: string;
  category: string;
  avatarUrl: string;
  coverUrl: string;
  membersCount: number;
  postsCount: number;
  isPrivate: boolean;
  isJoined: boolean;
  relevanceScore: number;
}

export interface SearchHashtagResult {
  hashtag: string;
  usageCount: number;
  lastUsed: string;
  recentPosts: string[];
  relevanceScore: number;
}

export interface SearchFilters {
  fromDate?: string;
  toDate?: string;
  categories?: string[];
  types?: string[];
  hasMedia?: boolean;
  minLikes?: number;
  minComments?: number;
  isVerified?: boolean;
  isPrivate?: boolean;
}

export interface SearchSort {
  field: string;
  direction: 'asc' | 'desc';
}

export interface TrendingSearch {
  query: string;
  searchCount: number;
  trendScore: number;
  lastSearched: string;
}

export interface SearchHistoryItem {
  id: string;
  query: string;
  resultCount: number;
  searchedAt: string;
}

export interface SearchAnalytics {
  totalSearches: number;
  uniqueQueries: number;
  mostSearchedTerms: SearchTermStats[];
  searchTrends: SearchTrend[];
  period: DateRange;
}

export interface SearchTermStats {
  term: string;
  count: number;
  trend: number;
}

export interface SearchTrend {
  date: string;
  searchCount: number;
  uniqueQueries: number;
}

export interface DateRange {
  from: string;
  to: string;
}

class SearchService extends BaseApiService {
  private searchHistory: SearchHistoryItem[] = [];
  private recentSearches: string[] = [];

  constructor() {
    super(socialApi);
    this.loadSearchHistory();
  }

  // Perform comprehensive search
  async search(request: SearchRequest): Promise<SearchResult> {
    try {
      const startTime = Date.now();
      
      // Perform searches in parallel for different types
      const searchPromises: Promise<any>[] = [];
      
      if (request.includeUsers !== false) {
        searchPromises.push(
          this.searchUsers(request.query, request.filters, request.sort)
            .catch(error => {
              console.error('User search failed:', error);
              return [];
            })
        );
      } else {
        searchPromises.push(Promise.resolve([]));
      }
      
      if (request.includePosts !== false) {
        searchPromises.push(
          this.searchPosts(request.query, request.filters, request.sort)
            .catch(error => {
              console.error('Post search failed:', error);
              return [];
            })
        );
      } else {
        searchPromises.push(Promise.resolve([]));
      }
      
      if (request.includeGroups !== false) {
        searchPromises.push(
          this.searchGroups(request.query, request.filters, request.sort)
            .catch(error => {
              console.error('Group search failed:', error);
              return [];
            })
        );
      } else {
        searchPromises.push(Promise.resolve([]));
      }
      
      if (request.includeHashtags !== false) {
        searchPromises.push(
          this.searchHashtags(request.query, request.sort)
            .catch(error => {
              console.error('Hashtag search failed:', error);
              return [];
            })
        );
      } else {
        searchPromises.push(Promise.resolve([]));
      }

      const [users, posts, groups, hashtags] = await Promise.all(searchPromises);
      
      const searchTime = Date.now() - startTime;
      
      return {
        query: request.query,
        totalResults: users.length + posts.length + groups.length + hashtags.length,
        posts: posts,
        users: users,
        groups: groups,
        hashtags: hashtags,
        suggestions: [], // Could be implemented later
        searchTime: `${searchTime}ms`
      };
    } catch (error) {
      console.error('Search failed:', error);
      throw error;
    }
  }

  // Search posts only
  async searchPosts(query: string, filters?: SearchFilters, sort?: SearchSort): Promise<SearchPostResult[]> {
    try {
      const params = new URLSearchParams();
      params.append('query', query);
      
      if (filters) {
        if (filters.fromDate) params.append('filters.fromDate', filters.fromDate);
        if (filters.toDate) params.append('filters.toDate', filters.toDate);
        if (filters.categories) params.append('filters.categories', filters.categories.join(','));
        if (filters.hasMedia !== undefined) params.append('filters.hasMedia', filters.hasMedia.toString());
        if (filters.minLikes) params.append('filters.minLikes', filters.minLikes.toString());
        if (filters.minComments) params.append('filters.minComments', filters.minComments.toString());
      }
      
      if (sort) {
        params.append('sort.field', sort.field);
        params.append('sort.direction', sort.direction);
      }

      const response = await this.get<{ data: SearchPostResult[] }>(`/search/posts?${params.toString()}`);
      return response.data;
    } catch (error) {
      console.error('Post search failed:', error);
      throw error;
    }
  }

  // Search users only
  async searchUsers(query: string, filters?: SearchFilters, sort?: SearchSort): Promise<SearchUserResult[]> {
    try {
      const params = new URLSearchParams();
      params.append('query', query);
      params.append('page', '1');
      params.append('limit', '20');
      
      if (filters) {
        if (filters.isVerified !== undefined) params.append('filters.isVerified', filters.isVerified.toString());
      }
      
      if (sort) {
        params.append('sort.field', sort.field);
        params.append('sort.direction', sort.direction);
      }

      // Use the social service follows endpoint for user search
      const response = await this.get<{ users: any[] }>(`/api/follows/search?${params.toString()}`);
      
      // Transform the response to match SearchUserResult interface
      return response.users.map((user: any) => ({
        id: user.id,
        username: user.username,
        displayName: user.displayName,
        bio: '', // Not provided by the API
        avatarUrl: user.avatarUrl || '',
        followersCount: user.followersCount || 0,
        followingCount: user.followingCount || 0,
        postsCount: user.postsCount || 0,
        isVerified: user.isVerified || false,
        isFollowing: false, // Not provided by the API
        relevanceScore: 1.0 // Default relevance score
      }));
    } catch (error) {
      console.error('User search failed:', error);
      throw error;
    }
  }

  // Search groups only
  async searchGroups(query: string, filters?: SearchFilters, sort?: SearchSort): Promise<SearchGroupResult[]> {
    try {
      const params = new URLSearchParams();
      params.append('query', query);
      
      if (filters) {
        if (filters.categories) params.append('filters.categories', filters.categories.join(','));
        if (filters.isPrivate !== undefined) params.append('filters.isPrivate', filters.isPrivate.toString());
      }
      
      if (sort) {
        params.append('sort.field', sort.field);
        params.append('sort.direction', sort.direction);
      }

      const response = await this.get<{ data: SearchGroupResult[] }>(`/search/groups?${params.toString()}`);
      return response.data;
    } catch (error) {
      console.error('Group search failed:', error);
      throw error;
    }
  }

  // Search hashtags only
  async searchHashtags(query: string, sort?: SearchSort): Promise<SearchHashtagResult[]> {
    try {
      const params = new URLSearchParams();
      params.append('query', query);
      
      if (sort) {
        params.append('sort.field', sort.field);
        params.append('sort.direction', sort.direction);
      }

      const response = await this.get<{ data: SearchHashtagResult[] }>(`/search/hashtags?${params.toString()}`);
      return response.data;
    } catch (error) {
      console.error('Hashtag search failed:', error);
      throw error;
    }
  }

  // Get search suggestions
  async getSearchSuggestions(query: string, limit: number = 10): Promise<string[]> {
    try {
      if (query.length < 2) {
        return [];
      }

      const params = new URLSearchParams();
      params.append('query', query);
      params.append('count', limit.toString());

      // Use neurosparkApi for search suggestions
      const response = await neurosparkApi.get<{ data: string[] }>(`/api/search/suggestions?${params.toString()}`);
      return response.data || [];
    } catch (error) {
      console.error('Failed to get search suggestions:', error);
      return [];
    }
  }

  // Get trending searches
  async getTrendingSearches(limit: number = 10): Promise<TrendingSearch[]> {
    try {
      const params = new URLSearchParams();
      params.append('limit', limit.toString());

      const response = await this.get<{ data: TrendingSearch[] }>(`/search/trending?${params.toString()}`);
      return response.data;
    } catch (error) {
      console.error('Failed to get trending searches:', error);
      return [];
    }
  }

  // Get search history
  async getSearchHistory(limit: number = 20): Promise<SearchHistoryItem[]> {
    try {
      const params = new URLSearchParams();
      params.append('limit', limit.toString());

      const response = await this.get<{ data: SearchHistoryItem[] }>(`/search/history?${params.toString()}`);
      this.searchHistory = response.data;
      return response.data;
    } catch (error) {
      console.error('Failed to get search history:', error);
      return this.searchHistory;
    }
  }

  // Clear search history
  async clearSearchHistory(): Promise<void> {
    try {
      await this.delete('/search/history');
      this.searchHistory = [];
      this.recentSearches = [];
      this.saveSearchHistory();
    } catch (error) {
      console.error('Failed to clear search history:', error);
    }
  }

  // Get search analytics (Admin only)
  async getSearchAnalytics(fromDate?: string, toDate?: string): Promise<SearchAnalytics> {
    try {
      const params = new URLSearchParams();
      if (fromDate) params.append('fromDate', fromDate);
      if (toDate) params.append('toDate', toDate);

      const response = await this.get<{ data: SearchAnalytics }>(`/search/analytics?${params.toString()}`);
      return response.data;
    } catch (error) {
      console.error('Failed to get search analytics:', error);
      throw error;
    }
  }

  // Advanced search
  async advancedSearch(request: SearchRequest): Promise<SearchResult> {
    try {
      const response = await this.post<{ data: SearchResult }>('/search/advanced', request);
      return response.data;
    } catch (error) {
      console.error('Advanced search failed:', error);
      throw error;
    }
  }

  // Utility methods
  parseSearchQuery(query: string): {
    terms: string[];
    hashtags: string[];
    mentions: string[];
    filters: Record<string, string>;
  } {
    const result = {
      terms: [] as string[],
      hashtags: [] as string[],
      mentions: [] as string[],
      filters: {} as Record<string, string>
    };

    if (!query) return result;

    // Extract hashtags
    const hashtagMatches = query.match(/#(\w+)/g);
    if (hashtagMatches) {
      result.hashtags = hashtagMatches.map(match => match.substring(1));
    }

    // Extract mentions
    const mentionMatches = query.match(/@(\w+)/g);
    if (mentionMatches) {
      result.mentions = mentionMatches.map(match => match.substring(1));
    }

    // Extract filters (e.g., "type:post", "user:john")
    const filterMatches = query.match(/(\w+):(\w+)/g);
    if (filterMatches) {
      filterMatches.forEach(match => {
        const [key, value] = match.split(':');
        result.filters[key] = value;
      });
    }

    // Extract remaining terms
    let cleanQuery = query;
    [...(hashtagMatches || []), ...(mentionMatches || []), ...(filterMatches || [])].forEach(match => {
      cleanQuery = cleanQuery.replace(match, '');
    });
    
    result.terms = cleanQuery.split(' ').filter(term => term.trim().length > 0);

    return result;
  }

  formatSearchQuery(parsed: ReturnType<typeof this.parseSearchQuery>): string {
    const parts: string[] = [];
    
    if (parsed.terms.length > 0) {
      parts.push(parsed.terms.join(' '));
    }
    
    if (parsed.hashtags.length > 0) {
      parts.push(parsed.hashtags.map(tag => `#${tag}`).join(' '));
    }
    
    if (parsed.mentions.length > 0) {
      parts.push(parsed.mentions.map(mention => `@${mention}`).join(' '));
    }
    
    if (Object.keys(parsed.filters).length > 0) {
      parts.push(Object.entries(parsed.filters).map(([key, value]) => `${key}:${value}`).join(' '));
    }

    return parts.join(' ');
  }

  addToRecentSearches(query: string): void {
    if (!query || query.trim().length === 0) return;
    
    // Remove if already exists
    this.recentSearches = this.recentSearches.filter(item => item !== query);
    
    // Add to beginning
    this.recentSearches.unshift(query);
    
    // Keep only last 10
    this.recentSearches = this.recentSearches.slice(0, 10);
    
    this.saveSearchHistory();
  }

  getRecentSearches(): string[] {
    return [...this.recentSearches];
  }

  clearRecentSearches(): void {
    this.recentSearches = [];
    this.saveSearchHistory();
  }

  private loadSearchHistory(): void {
    try {
      const saved = localStorage.getItem('searchHistory');
      if (saved) {
        const data = JSON.parse(saved);
        this.recentSearches = data.recentSearches || [];
        this.searchHistory = data.searchHistory || [];
      }
    } catch (error) {
      console.error('Failed to load search history:', error);
    }
  }

  private saveSearchHistory(): void {
    try {
      const data = {
        recentSearches: this.recentSearches,
        searchHistory: this.searchHistory
      };
      localStorage.setItem('searchHistory', JSON.stringify(data));
    } catch (error) {
      console.error('Failed to save search history:', error);
    }
  }

  // Search result formatting utilities
  formatSearchResultCount(count: number): string {
    if (count === 0) return 'No results';
    if (count === 1) return '1 result';
    if (count < 1000) return `${count} results`;
    if (count < 1000000) return `${(count / 1000).toFixed(1)}K results`;
    return `${(count / 1000000).toFixed(1)}M results`;
  }

  getSearchResultIcon(type: 'posts' | 'users' | 'groups' | 'hashtags'): string {
    const icons = {
      posts: '📝',
      users: '👤',
      groups: '👥',
      hashtags: '#️⃣'
    };
    return icons[type];
  }

  getSearchResultColor(type: 'posts' | 'users' | 'groups' | 'hashtags'): string {
    const colors = {
      posts: 'text-blue-600',
      users: 'text-green-600',
      groups: 'text-purple-600',
      hashtags: 'text-orange-600'
    };
    return colors[type];
  }

  highlightSearchTerms(text: string, query: string): string {
    if (!query || !text) return text;
    
    const terms = query.split(' ').filter(term => term.length > 0);
    let highlightedText = text;
    
    terms.forEach(term => {
      const regex = new RegExp(`(${term})`, 'gi');
      highlightedText = highlightedText.replace(regex, '<mark class="bg-yellow-200">$1</mark>');
    });
    
    return highlightedText;
  }
}

export const searchService = new SearchService();
